cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/BusinessMetadataDataset.java

```diff
@@ -0,0 +1,151 @@
+    write(mdsKey, metadataRecord);
+  }
+
+  /**
+   * Add new business metadata.
+   *
+   * @param targetId The target Id: app-id(ns+app) / program-id(ns+app+pgtype+pgm) /
+   *                 dataset-id(ns+dataset)/stream-id(ns+stream).
+   * @param key The metadata key to be added.
+   * @param value The metadata value to be added.
+   */
+  public void createBusinessMetadata(Id.NamespacedId targetId, String key, String value) {
+    createBusinessMetadata(new BusinessMetadataRecord(getTargetType(targetId), targetId, key, value));
+  }
+
+  /**
+   * Return business metadata based on type, target id, and key.
+   *
+   * @param targetId The id of the target.
+   * @param key The metadata key to get.
+   * @return instance of {@link BusinessMetadataRecord} for the target type, id, and key.
+   */
+  public BusinessMetadataRecord getBusinessMetadata(Id.NamespacedId targetId, String key) {
+    String targetType = getTargetType(targetId);
+    MDSKey mdsKey = getInstanceKey(targetType, targetId, key);
+    Row row = indexedTable.get(mdsKey.getKey());
+    if (row.isEmpty()) {
+      return null;
+    }
+
+    byte[] keyvalue = row.get(KEYVALUE_COLUMN);
+    byte[] value = row.get(VALUE_COLUMN);
+
+    return new BusinessMetadataRecord(targetType, targetId, key, Bytes.toString(value));
+  }
+
+  private static Id.NamespacedId fromString(String type, String id) {
+    if (type.equals(Id.Program.class.getSimpleName())) {
+      return Id.Program.fromStrings(id.split("/"));
+    } else if (type.equals(Id.Application.class.getSimpleName())) {
+      return Id.Application.fromStrings(id.split("/"));
+    } else if (type.equals(Id.DatasetInstance.class.getSimpleName())) {
+      // TODO Add code
+    } else if (type.equals(Id.Stream.class.getSimpleName())) {
+      return Id.Stream.fromId(id);
+    }
+    throw new IllegalArgumentException("Illegal Type of metadata source.");
+  }
+
+  /**
+   * Find the instance of {@link BusinessMetadataRecord} based on key.
+   *
+   * @param key The metadata value to be found.
+   * @return Collection of {@link BusinessMetadataRecord} fits the key.
+   */
+  public Collection<BusinessMetadataRecord> findBusinessMetadataOnKey(String key) {
+
+    // TODO ADD CODE
+
+    return Lists.newArrayList();
+  }
+
+  private void write(MDSKey id, BusinessMetadataRecord record) {
+    try {
+      Put put = new Put(id.getKey());
+
+      // Now add the index columns.
+      put.add(Bytes.toBytes(KEYVALUE_COLUMN), Bytes.toBytes(record.getKey() + ":" + record.getValue()));
+      put.add(Bytes.toBytes(VALUE_COLUMN), Bytes.toBytes(record.getValue()));
+
+      indexedTable.put(put);
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  // Helper method to generate key.
+  private MDSKey getInstanceKey(String targetType, Id.NamespacedId targetId, String key) {
+    MDSKey.Builder builder = new MDSKey.Builder();
+    builder.add(targetType);
+    builder.add(targetId.toString());
```

Comment:
I don't think `toString` methods of `Id` are reliable, we should probably do our own serialization/de-serialization here.